k8srm-prototype/pkg/api/capacity_types.go

```diff
@@ -40,9 +40,17 @@ type DevicePoolSpec struct {
 	// +optional
 	Attributes []Attribute `json:"attributes,omitempty"`
 
-	// DeviceCount contains the total number of devices in the pool.
+	// Resources are pooled resources that are shared by all devices in the
+	// pool. This is typically used when representing a partitionable
+	// device, and need not be populated otherwise.
+	//
+	// +optional
+	Resources []ResourceCapacity `json:"resources,omitempty"`
```

Comment:
This has lost one level of indirection from what my prototype had. Mine had a list of named groups (each of which had a list of named capacities), whereas this is just a list of named capaacities. This implies, I will need one DevicePool per GPU (rather than a single device pool for all GPUs).

Reply:
Yes. Trade off for simpler model. Worth discussing.

Reply:
Here's an interesting thought where your indirection *might* be useful (thinking aloud here).

We have this use case of "give me a GPU and NIC and make sure they are *close*". Right now, we have the user solve that in their claim: they can request two different device in the same claim, and they can put `matchAttributes` to align the topologies.

That's a lot to put on the user. Another approach @thockin and I have discussed is a "compound device". We could combine the two devices into one on the model side, so that "you get the GPU, you automatically get the NIC" (and only the GPU would be advertised). Classic DRA solves this with a special compound device driver that understands the subordinate devices. I don't think that's the right approach, as it leads to a lot of interdependencies.

One thought we had was some sort of hook in the driver framework, that lets us inject stuff into the driver's model. The idea is that the cloud vendor designing the node shapes would know "GPU 1 goes with NIC 1" etc., and could encode this and inject it into the well-known device model.

But where to inject it? If we allowed multiple shared resource groups, we could create a group of NICs in the pool, and add a request to the individual devices.

As I write that out, I am not sure it's a great idea. But bears some thinking and exploring.

Reply:
A question that was asked which I found interesting:

Today, the workloads we see most are expressing "give me a whole machine (and many of them) connected by fast network".  We (I) believe thast eventually we will need to answer "give me less than the whole machine".  For those cases - do they really need the NIC?  If they ask for 2 GPUs with 2 replicas, that will almost certainly perform worse than asking for 4 GPUs.

Why would someone actually ask for "less than the whole machine, but many replicas of that"?  Are we complicating things for a use-case nobody really has?

Reply:
What do you mean by „replicas“ here? We have the notion of „replicas“ in the existing GPU device plugin to oversubscribed access to individual GPUs, but I never plan on implementing anything like that in our DRA driver.

Each GPU that gets allocated will be a distinct physical GPU. Any sharing / oversubscribing that then happens will occur because two workloads have opted in to point to the same claim that has a given GPU (or set of GPUs) bound to it,

Reply:
A representative use case in asked about often is the last one in my doc which I still have no idea how to solve with structure parameters: https://docs.google.com/document/u/0/d/1bDO11rEq_Yhpgpk5RN0VwnMLV1_2wNWvtOyM_QoIV_Y/mobilebasic?pli=1#heading=h.5sgo286txzv5

That is:

Requirements:
* Container A in Pod A needs 10GB of GPU memory
* Container B in Pod B needs 12GB of GPU memory


Desired Allocation Logic:
* If a GPU with >22GB of memory is available, run both containers time-sliced on one GPU
* Elseif two GPUs are present on the same node with >=12 GB of memory each, run them on neighboring GPUs on the same node
* Elseif two GPUs are present with >=12 GB of memory each on separate nodes and each have an RDMA capable NIC, run them on those GPUs
* Else throw an error

Reply:
I think you almost answered my question.  You describe 2 distinct pods (not replicas) that each want less than "a whole machine" and which would be OK to land on 2 different nodes, as long as they have fast network.  They would be BETTER on one node with NVLink, but will accept 2 nodes with RDMA.

Is that right?

So here's the part of my question I don't think you answered:  Is this REAL or hypothetical?

Reply:
Two distinct pods - meaning they are handcrafted by the user? Is there some workload controller here? Seems a strange thing to do.

Reply:
> * If a GPU with >22GB of memory is available, run both containers time-sliced on one GPU
> * Elseif two GPUs are present on the same node with >=12 GB of memory each, run them on neighboring GPUs on the same node
> * Elseif two GPUs are present with >=12 GB of memory each on separate nodes and each have an RDMA capable NIC, run them on those GPUs
> * Else throw an error

The first two are do-able in this model. The last one is not. That is true in classic DRA as well as structured parameters, I think. In all cases we cannot split a single claim across nodes. I think cross-node things are out-of-scope at the DRA level.

I can see possible ways of eventually having a multi-host claim, and node becomes a `matchAttributes` with a "preferred" rather than "required" level of constraint. But I didn't think we would go there this year...

Reply:
Actually maybe we can't, even do the first two. Let me think on it. But how real is this? Why separate pods rather than one pod and multiple containers? We may need a higher level construct to solve this one.

Reply:
The ability to do exactly this (with these exact numbers) is what got NVIDIA excited about DRA in the first place. This is a real application that has been struggling to provide a cost effective solution for years.

Reply:
To be clear, I’m not suggesting we solve this use case with structured parameters, but it does suggest we shouldn’t rip out the ability to do „classic“ DRA for more advanced use cases.

Reply:
How did you do multi-node things in classic DRA?

Reply:
Or, maybe, more accurately, given that classic DRA is still a "schedule one pod at a time", how are you doing any multi-pod orchestration? I can see manual association of pods with a claim could be done. But the logic you presented would seem to require considering both pods in concert, rather than one at a time.

Reply:
We have a custom ClaimParameters object which allows us to chain multiple, dependent claims together as well as encode the semantics defined in: https://github.com/kubernetes-sigs/wg-device-management/pull/5#discussion_r1589687820

* One pod references a claim with claim parameters asking for 10GB of GPU Memory
* A second pod references a claim with claim parameters asking for 12GB of GPU Memory
* Each of these claim parameters have a common field which "links" them together (similar to match attributes)

Regardless of which pod gets scheduled first, the allocation of all "linked" claims gets done together, thus reserving the necessary GPUs for all pods referencing these linked claims. As the encoded semantics imply, these allocations will sometimes be on the same GPU, sometimes on the same node, and other times across two different nodes with an additional RDMA interface being injected.

Of course this implies that subsequent pods better be able to land on the nodes where the GPUs they require have been pre-allocated, but at least for the use-case this supports at NVIDIA, this will always be the case.

Reply:
I wouldn't necessarily call them "sub-claims" but rather "sibling-claims". And it's not _technically_ immediate allocation because none of them get allocated until the first consumer of the first claim comes along. But the on-demand allocation of the first claim triggers the immediate allocation of its sibling-claims.